docs/proposals/pod-resource-management.md

```diff
@@ -0,0 +1,394 @@
+
+```
+/memory/G-PodUID.slice/containerUID.scope
+/cpu,cpuacct/G-PodUID.slice/containerUID.scope
+/memory/Bu.slice/Bu-PodUID.slice/containerUID.scope
+/cpu,cpuacct/Bu.slice/Bu-PodUID.slice/containerUID.scope
+/memory/BE.slice/BE-PodUID.slice/containerUID.scope
+/cpu,cpuacct/BE.slice/BE-PodUID.slice/containerUID.scope
+```
+
+### Hierarchy Outline
+
+- "$Root" is the system root of the node ie. "/" by default and if `--cgroup-root` is specified then the specified cgroup-root is used as "$Root".
+- We have a top level Qos cgroup for the `Bu`(Burstable) and `BE`(Best Effort) Qos classes.
+- But we __dont__ have a separate cgroup for the `G`(Guaranteed) QoS class. Guaranteed pod cgroups are brought up directly under the `Root` cgroup.
+- Each pod has its own cgroup which is nested under the cgroup matching the pod's Qos class.
+- All containers brought up by the pod are nested under the pod's cgroup.
+- system-reserved cgroup contains the system specific processes.
+- kube-reserved cgroup contains the kubelet specific daemons.
+
+```
+$ROOT
+  |
+  +- Pod1
+  |   |
+  |   +- Container1
+  |   +- Container2
+  |   ...
+  +- Pod2
+  |   +- Container3
+  |   ...
+  +- ...
+  |
+  +- Bu
+  |   |
+  |   +- Pod3
+  |   |   |
+  |   |   +- Container4
+  |   |   ...
+  |   +- Pod4
+  |   |   +- Container5
+  |   |   ...
+  |   +- ...
+  |
+  +- BE
+  |   |
+  |   +- Pod5
+  |   |   |
+  |   |   +- Container6
+  |   |   +- Container7
+  |   |   ...
+  |   +- ...
+  |
+  +- System-reserved
+  |   |
+  |   +- system
+  |   +- ...
+  |
+  +- Kube-reserved 
+  |   |
+  |   +- kubelet
+  |   +- docker (optional)
+  |   +- ...
+  |
+```
+
+QoS Policy Design Decisions
+
+- This hierarchy highly prioritizes resource guarantees to the Guaranteed over Burstable and Best Effort Pods.
+- By not having a separate cgroup for the Guaranteed class, the hierarchy allows the Guaranteed pods to burst and utilize all of Node's Allocatable capacity.
+- The Best Effort and Burstable pods are strictly restricted from bursting and hogging resources and thus `G` Pods are guaranteed resource isolation.
+- Best Effort pods are treated as lowest priority. So for the `BE` QoS cgroup we set cpu shares to the lowest possible value ie.2. This ensures that the best effort containers get a relatively small share of cpu time.
+- Also we don't set any quota on the cpu resources as the containers on the `BE` pods can use any amount of free resources on the node.
+- Having memory limit of `BE` cgroup as (Allocatable - summation of memory limits of Guaranteed and Burstable pods) would result in `BE` pods becoming more susceptible to being OOM killed. As more `G` and `Bu` pods are scheduled kubelet will more likely kill `BE` pods, even if the `G` and `Bu` pods are using less than their request since we will be dynamically reducing the size of `BE` m.limit_in_bytes. But this allows for better memory guarantees to the `G` and `Bu` pods.
```

Comment:
And if the goal is to avoid at all costs the system OOM killer being invoked, it seems like you could still have something like @derekwaynecarr 's evictor evict when the machine gets close to OOMing based on actual  memory usage.